src/network/dn_expand.c

```diff
@@ -1,28 +1,35 @@
 #include <resolv.h>
 
-int __dn_expand(const unsigned char *base, const unsigned char *end, const unsigned char *src, char *dest, int space)
+_Checked int __dn_expand(const unsigned char *base : bounds(base, end) itype(_Array_ptr<const unsigned char>),
+	const unsigned char *end : itype(_Ptr<const unsigned char>),
+	const unsigned char *src : bounds(src, end) itype(_Array_ptr<const unsigned char>),
+	char *dest : count(space > 254 ? 254 : space) itype(_Array_ptr<char>),
+	int space)
 {
-	const unsigned char *p = src;
-	char *dend, *dbegin = dest;
+	const int space0 = space > 254 ? 254 : space;
+	_Array_ptr<const unsigned char> p : bounds(src, end) = src;
 	int len = -1, i, j;
 	if (p==end || space <= 0) return -1;
-	dend = dest + (space > 254 ? 254 : space);
+	const _Array_ptr<const char> dbegin : count(space0) = dest;
+	const _Array_ptr<const char> dend : count(0) = dest + space0;
+	// dst loops through bounds(dbegin, dend).
+	_Array_ptr<char> dst : bounds(dbegin, dend) = dest;
 	/* detect reference loop using an iteration counter */
 	for (i=0; i < end-base; i+=2) {
-		/* loop invariants: p<end, dest<dend */
+		/* loop invariants: p<end, dst<dend */
 		if (*p & 0xc0) {
 			if (p+1==end) return -1;
 			j = ((p[0] & 0x3f) << 8) | p[1];
 			if (len < 0) len = p+2-src;
 			if (j >= end-base) return -1;
 			p = base+j;
 		} else if (*p) {
-			if (dest != dbegin) *dest++ = '.';
+			if (dst != dbegin) *dst++ = '.';
 			j = *p++;
-			if (j >= end-p || j >= dend-dest) return -1;
-			while (j--) *dest++ = *p++;
+			if (j >= end-p || j >= dend-dst) return -1;
+			while (j--) *dst++ = *p++;
```

Comment:
Don't you need to break `*dst++` into `*dst` followed by `dst++` because of the runtime error that you reported?

Reply:
`dst` does not use self-referential bounds. The undefined behavior in https://github.com/microsoft/checkedc-clang/issues/863 involves incrementing a pointer with a self-referential bound.

Reply:
Can you add a clarifying comment to https://github.com/microsoft/checkedc-clang/issues/863 that the issue is only for pointers with self-referential bounds?

Reply:
Added. Thanks!